server/service/src/processors/transfer/invoice/create_inbound_invoice.rs

```diff
@@ -60,40 +65,80 @@ impl InvoiceTransferProcessor for CreateInboundInvoiceProcessor {
+                operation => {
+                    return Ok(InvoiceTransferOutput::WrongOperation(operation.to_owned()))
+                }
             };
+
         // 2.
         // Also get type for new invoice
-        let new_invoice_type = match outbound_invoice.invoice_row.r#type {
+        let new_invoice_type = match &outbound_invoice.invoice_row.r#type {
             InvoiceType::OutboundShipment => InboundInvoiceType::InboundShipment,
             InvoiceType::SupplierReturn => InboundInvoiceType::CustomerReturn,
-            _ => return Ok(None),
+            other => return Ok(InvoiceTransferOutput::WrongType(other.to_owned())),
         };
 
         // 3.
         if !matches!(
-            outbound_invoice.invoice_row.status,
+            &outbound_invoice.invoice_row.status,
             InvoiceStatus::Shipped | InvoiceStatus::Picked
         ) {
-            return Ok(None);
+            return Ok(InvoiceTransferOutput::WrongOutboundStatus(
+                outbound_invoice.invoice_row.status.to_owned(),
+            ));
         }
+
         // 4.
         if linked_invoice.is_some() {
-            return Ok(None);
+            return Ok(InvoiceTransferOutput::NoLinkedInvoice);
         }
+
         // 5.
         let store = StoreRowRepository::new(connection)
             .find_one_by_id(&record_for_processing.other_party_store_id)?
             .ok_or(RepositoryError::NotFound)?;
-
         if let Some(created_date) = store.created_date {
-            let store_created_datetime = NaiveDateTime::new(
-                created_date - Duration::days(30),
-                NaiveTime::from_hms_opt(0, 0, 0).unwrap_or_default(),
-            );
+            let store_created_datetime =
+                NaiveDateTime::new(created_date - Duration::days(30), Default::default());
             let invoice_created_datetime = outbound_invoice.invoice_row.created_datetime;
             if invoice_created_datetime < store_created_datetime {
-                return Ok(None);
+                return Ok(InvoiceTransferOutput::InvoiceCreatedBeforeStore);
+            }
+        }
+
+        // 6.
+        if outbound_invoice.invoice_row.status == InvoiceStatus::Picked {
+            if let Some(picked_date) = outbound_invoice.invoice_row.picked_datetime {
+                let pref_months = PreventTransfersMonthsBeforeInitialisation {}
+                    .load(connection, None)
+                    .map_err(|e| RepositoryError::DBError {
+                        msg: e.to_string(),
+                        extra: "".to_string(),
+                    })?;
+                if pref_months > 0 {
+                    let sort = Sort {
+                        key: SyncLogSortField::DoneDatetime,
+                        desc: None,
+                    };
+
+                    let latest_log_sorted_by_finished_datetime = SyncLogRepository::new(connection)
+                        .query(Pagination::one(), None, Some(sort))?
```

Comment:
First log might not have integration_finished_datetime I think, should filter by that one being 'not null'.

We also may decide to trim here, I think we should look at new key_value 'value' with initialisation date <- but can do that in develop

Reply:
Yea you're right the nice

Reply:
Turns out when sorted ascending, nulls are at the end of the sql result rather than start so this pagination/pop ensures that we get the earliest log. Regardless I've added a filter to be safe!